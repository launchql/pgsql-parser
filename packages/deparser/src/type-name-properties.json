{
  "TypeCast": ["typeName"],
  "RangeTableFuncCol": ["typeName"],
  "ColumnDef": ["typeName"],
  "XmlSerialize": ["typeName"],
  "JsonOutput": ["typeName"],
  "JsonTableColumn": ["typeName"],
  "CreateDomainStmt": ["typeName"],
  "CreateTransformStmt": ["type_name"],
  "CreateStmt": ["ofTypename"],
  "CreateOpClassItem": ["datatype", "storedtype"],
  "CreateFunctionStmt": ["returnType"],
  "FunctionParameter": ["argType"],
  "CreateCastStmt": ["sourcetype", "targettype"]
} 