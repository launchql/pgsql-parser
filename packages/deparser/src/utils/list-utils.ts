import { Node } from '@pgsql/types';

export class ListUtils {
  static unwrapList(obj: any): any[] {
    if (obj === undefined || obj === null) {
      return [];
    }
    
    if (obj.List !== undefined) {
      return obj.List.items || [];
    }
    
    if (Array.isArray(obj)) {
      return obj;
    }
    
    return [obj];
  }

  static formatList(items: any[], separator: string = ', ', prefix: string = '', formatter: (item: any) => string): string {
    if (!items || items.length === 0) {
      return '';
    }

    return items
      .map(item => `${prefix}${formatter(item)}`)
      .join(separator);
  }
}
