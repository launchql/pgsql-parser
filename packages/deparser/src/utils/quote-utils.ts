const RESERVED_WORDS = new Set([
  'all', 'analyse', 'analyze', 'and', 'any', 'array', 'as', 'asc', 'asymmetric',
  'authorization', 'binary', 'both', 'case', 'cast', 'check', 'collate', 'collation',
  'column', 'concurrently', 'constraint', 'create', 'cross', 'current_catalog',
  'current_date', 'current_role', 'current_schema', 'current_time', 'current_timestamp',
  'current_user', 'default', 'deferrable', 'desc', 'distinct', 'do', 'else', 'end',
  'except', 'false', 'fetch', 'for', 'foreign', 'freeze', 'from', 'full', 'grant',
  'group', 'having', 'ilike', 'in', 'initially', 'inner', 'intersect', 'into', 'is',
  'isnull', 'join', 'lateral', 'leading', 'left', 'like', 'limit', 'localtime',
  'localtimestamp', 'natural', 'not', 'notnull', 'null', 'offset', 'on', 'only',
  'or', 'order', 'outer', 'overlaps', 'placing', 'primary', 'references', 'returning',
  'right', 'select', 'session_user', 'similar', 'some', 'symmetric', 'table', 'tablesample',
  'then', 'to', 'trailing', 'true', 'union', 'unique', 'user', 'using', 'variadic',
  'verbose', 'when', 'where', 'window', 'with'
]);

export class QuoteUtils {
  static needsQuotes(value: string): boolean {
    if (!value || typeof value !== 'string') {
      return false;
    }

    const lowerValue = value.toLowerCase();
    
    if (RESERVED_WORDS.has(lowerValue)) {
      return true;
    }

    if (!/^[a-z_][a-z0-9_$]*$/i.test(value)) {
      return true;
    }

    if (value !== value.toLowerCase()) {
      return true;
    }

    return false;
  }

  static quote(value: any): any {
    if (value == null) {
      return null;
    }

    if (Array.isArray(value)) {
      return value.map(v => this.quote(v));
    }

    if (typeof value !== 'string') {
      return value;
    }

    if (this.needsQuotes(value)) {
      return `"${value}"`;
    }

    return value;
  }

  static escape(literal: string): string {
    return `'${literal.replace(/'/g, "''")}'`;
  }
}
