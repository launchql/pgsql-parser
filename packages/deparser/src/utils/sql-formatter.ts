export class SqlFormatter {
  private newlineChar: string;
  private tabChar: string;
  private prettyMode: boolean;

  constructor(newlineChar: string = '\n', tabChar: string = '  ', prettyMode: boolean = true) {
    this.newlineChar = newlineChar;
    this.tabChar = tabChar;
    this.prettyMode = prettyMode;
  }

  format(parts: string[], separator: string = ' '): string {
    return parts.filter(part => part !== null && part !== undefined && part !== '').join(separator);
  }

  indent(text: string, count: number = 1): string {
    if (!this.prettyMode) {
      return text;
    }
    const indentation = this.tabChar.repeat(count);
    return text.split(this.newlineChar).map(line => 
      line.trim() ? indentation + line : line
    ).join(this.newlineChar);
  }

  parens(content: string): string {
    return `(${content})`;
  }

  newline(): string {
    return this.newlineChar;
  }

  tab(): string {
    return this.tabChar;
  }

  isPretty(): boolean {
    return this.prettyMode;
  }
}
