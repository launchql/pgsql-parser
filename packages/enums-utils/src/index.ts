import toInt from './enums2int.json';
import toStr from './enums2str.json';
import nodes from './nodes.json';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
