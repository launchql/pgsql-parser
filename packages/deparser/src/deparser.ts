import { Node } from '@pgsql/types';
import { SqlFormatter } from './utils/sql-formatter';
import { ConsolidatedVisitor } from './visitors/consolidated-visitor';
import { DeparserContext } from './visitors/base';

export interface DeparserOptions {
  newline?: string;
  tab?: string;
}

export class Deparser {
  private formatter: SqlFormatter;
  private consolidatedVisitor: ConsolidatedVisitor;
  private tree: Node[];

  constructor(tree: Node | Node[], opts: DeparserOptions = {}) {
    this.formatter = new SqlFormatter(opts.newline, opts.tab);
    this.consolidatedVisitor = new ConsolidatedVisitor(this.formatter, this);
    
    this.tree = Array.isArray(tree) ? tree : [tree];
  }

  static deparse(query: Node | Node[], opts: DeparserOptions = {}): string {
    return new Deparser(query, opts).deparseQuery();
  }

  deparseQuery(): string {
    return this.tree
      .map(node => this.deparse(node))
      .join(this.formatter.newline() + this.formatter.newline());
  }

  deparse(node: Node, context: DeparserContext = {}): string | null {
    if (node == null) {
      return null;
    }

    if (typeof node === 'number' || node instanceof Number) {
      return node.toString();
    }

    try {
      return this.consolidatedVisitor.visit(node, context);
    } catch (error) {
      const nodeType = Object.keys(node)[0];
      throw new Error(`Error deparsing ${nodeType}: ${(error as Error).message}`);
    }
  }


}
