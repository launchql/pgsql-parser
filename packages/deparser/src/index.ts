import { Deparser } from "./deparser";

const deparse = Deparser.deparse;
export { deparse, Deparser };
