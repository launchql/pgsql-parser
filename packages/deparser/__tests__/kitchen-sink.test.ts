import { parse } from '@pgsql/parser';
import { deparse } from '../src';
import { cleanTree, cleanLines } from '../src/utils';
import { readFileSync } from 'fs';
import * as path from 'path';

const GENERATED_JSON = path.join(__dirname, '../../../__fixtures__/generated/generated.json');
const SINGLE_TEST = process.env.SINGLE_TEST;
// const SINGLE_TEST = 'upstream/rowtypes-88.sql';

function printErrorMessage(sql: string, position: number) {
  const lineNumber = sql.slice(0, position).match(/\n/g)?.length || 0;
  const lines = sql.split('\n');
  let colNumber = position - 1;
  for (let l = 0; l < lineNumber; l++) {
    colNumber -= lines[l].length + 1;
  }
  const errMessage = [`Error line ${lineNumber + 1}, column ${colNumber + 1}`];
  // print the previous line if there is one
  if (lineNumber > 0) {
    errMessage.push(lines[lineNumber - 1]);
  }
  // print the errant line itself
  errMessage.push(lines[lineNumber]);
  // print a marker at the exact column
  errMessage.push(' '.repeat(colNumber) + '^');
  // print the line after if there is one
  if (lineNumber < lines.length - 1) {
    errMessage.push(lines[lineNumber + 1]);
  }
  console.error(errMessage.join('\n'));
}
function tryParse(sql: string) {
  try {
    return parse(sql);
  } catch (err: any) {
    // if there is position information, print a readable error message
    if (err.cursorPosition) {
      printErrorMessage(sql, err.cursorPosition);
    }
    throw err;
  }
}

// Read aggregated fixtures from JSON
const fixtures: Record<string, string> = JSON.parse(readFileSync(GENERATED_JSON, 'utf-8'));

describe('kitchen sink', () => {
  // Determine which entries to test: if SINGLE_TEST is set, filter to match; otherwise test all
  const entries = SINGLE_TEST
    ? Object.entries(fixtures).filter(([relPath]) => relPath === SINGLE_TEST)
    : Object.entries(fixtures);

  entries.forEach(([relativePath, sql]) => {
    it(relativePath, () => {
      const tree = tryParse(sql);
      if (tree.stmts) {
        tree.stmts.forEach((stmt) => {
          expect(stmt).toMatchSnapshot();
          if (stmt.stmt) {
            const outSql = deparse(stmt.stmt);

            expect(cleanLines(outSql)).toMatchSnapshot();
            const reparsed = parse(outSql);
            expect(cleanTree(reparsed.stmts || [])).toEqual(cleanTree([stmt]));
          }
        });
      }
    });
  });
});
