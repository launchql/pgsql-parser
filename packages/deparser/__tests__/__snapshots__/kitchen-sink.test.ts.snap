// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kitchen sink privs-and-defaults-1.sql 1`] = `
{
  "stmt": {
    "AlterDefaultPrivilegesStmt": {
      "action": {
        "behavior": "DROP_RESTRICT",
        "grantees": [
          {
            "RoleSpec": {
              "location": 73,
              "rolename": "your_user",
              "roletype": "ROLESPEC_CSTRING",
            },
          },
        ],
        "is_grant": true,
        "objtype": "OBJECT_TABLE",
        "privileges": [
          {
            "AccessPriv": {
              "priv_name": "select",
            },
          },
        ],
        "targtype": "ACL_TARGET_DEFAULTS",
      },
      "options": [
        {
          "DefElem": {
            "arg": {
              "List": {
                "items": [
                  {
                    "String": {
                      "sval": "schema_name",
                    },
                  },
                ],
              },
            },
            "defaction": "DEFELEM_UNSPEC",
            "defname": "schemas",
            "location": 25,
          },
        },
      ],
    },
  },
}
`;

exports[`kitchen sink privs-and-defaults-1.sql 2`] = `"ALTER DEFAULT PRIVILEGES IN SCHEMA schema_name GRANT SELECT ON TABLES TO your_user"`;

exports[`kitchen sink privs-and-defaults-2.sql 1`] = `
{
  "stmt": {
    "AlterDefaultPrivilegesStmt": {
      "action": {
        "behavior": "DROP_RESTRICT",
        "grantees": [
          {
            "RoleSpec": {
              "location": 70,
              "rolename": "your_user",
              "roletype": "ROLESPEC_CSTRING",
            },
          },
        ],
        "is_grant": true,
        "objtype": "OBJECT_TABLE",
        "targtype": "ACL_TARGET_DEFAULTS",
      },
      "options": [
        {
          "DefElem": {
            "arg": {
              "List": {
                "items": [
                  {
                    "String": {
                      "sval": "schema_name",
                    },
                  },
                ],
              },
            },
            "defaction": "DEFELEM_UNSPEC",
            "defname": "schemas",
            "location": 25,
          },
        },
      ],
    },
  },
}
`;

exports[`kitchen sink privs-and-defaults-2.sql 2`] = `"ALTER DEFAULT PRIVILEGES IN SCHEMA schema_name GRANT ALL PRIVILEGES ON TABLES TO your_user"`;

exports[`kitchen sink privs-and-defaults-3.sql 1`] = `
{
  "stmt": {
    "AlterDefaultPrivilegesStmt": {
      "action": {
        "behavior": "DROP_RESTRICT",
        "grantees": [
          {
            "RoleSpec": {
              "location": 73,
              "rolename": "your_user",
              "roletype": "ROLESPEC_CSTRING",
            },
          },
        ],
        "is_grant": true,
        "objtype": "OBJECT_SEQUENCE",
        "targtype": "ACL_TARGET_DEFAULTS",
      },
      "options": [
        {
          "DefElem": {
            "arg": {
              "List": {
                "items": [
                  {
                    "String": {
                      "sval": "schema_name",
                    },
                  },
                ],
              },
            },
            "defaction": "DEFELEM_UNSPEC",
            "defname": "schemas",
            "location": 25,
          },
        },
      ],
    },
  },
}
`;

exports[`kitchen sink privs-and-defaults-3.sql 2`] = `"ALTER DEFAULT PRIVILEGES IN SCHEMA schema_name GRANT ALL PRIVILEGES ON SEQUENCES TO your_user"`;

exports[`kitchen sink privs-and-defaults-4.sql 1`] = `
{
  "stmt": {
    "AlterDefaultPrivilegesStmt": {
      "action": {
        "behavior": "DROP_RESTRICT",
        "grantees": [
          {
            "RoleSpec": {
              "location": 73,
              "rolename": "your_user",
              "roletype": "ROLESPEC_CSTRING",
            },
          },
        ],
        "is_grant": true,
        "objtype": "OBJECT_FUNCTION",
        "targtype": "ACL_TARGET_DEFAULTS",
      },
      "options": [
        {
          "DefElem": {
            "arg": {
              "List": {
                "items": [
                  {
                    "String": {
                      "sval": "schema_name",
                    },
                  },
                ],
              },
            },
            "defaction": "DEFELEM_UNSPEC",
            "defname": "schemas",
            "location": 25,
          },
        },
      ],
    },
  },
}
`;

exports[`kitchen sink privs-and-defaults-4.sql 2`] = `"ALTER DEFAULT PRIVILEGES IN SCHEMA schema_name GRANT ALL PRIVILEGES ON FUNCTIONS TO your_user"`;
